File: Frontend/src/services/feeService.ts
import api from './api';

export interface Fee {
  id: number;
  student: {
    id: number;
    firstName: string;
    lastName: string;
    email: string;
  };
  feeType: 'TUITION' | 'LIBRARY' | 'LABORATORY' | 'HOSTEL' | 'EXAMINATION' | 'MISCELLANEOUS';
  amount: number;
  paidAmount: number;
  dueDate?: string;
  paidDate?: string;
  paymentStatus: 'PENDING' | 'PARTIAL' | 'PAID' | 'OVERDUE';
  semester?: string;
  academicYear?: string;
  description?: string;
  createdAt: string;
  updatedAt: string;
}

export interface FeeRequest {
  studentId: number;
  feeType: 'TUITION' | 'LIBRARY' | 'LABORATORY' | 'HOSTEL' | 'EXAMINATION' | 'MISCELLANEOUS';
  amount: number;
  dueDate?: string;
  semester?: string;
  academicYear?: string;
  description?: string;
}

export interface FeesSummary {
  totalFees: number;
  totalPaid: number;
  outstanding: number;
}

export const feeService = {
  createFee: async (feeData: FeeRequest): Promise<Fee> => {
    const response = await api.post('/fees', feeData);
    return response.data;
  },

  getAllFees: async (): Promise<Fee[]> => {
    const response = await api.get('/fees');
    return response.data;
  },

  getFeeById: async (id: number): Promise<Fee> => {
    const response = await api.get(`/fees/${id}`);
    return response.data;
  },

  getStudentFees: async (studentId: number): Promise<Fee[]> => {
    const response = await api.get(`/fees/student/${studentId}`);
    return response.data;
  },

  payFee: async (feeId: number, amount: number): Promise<Fee> => {
    const response = await api.post(`/fees/${feeId}/pay?amount=${amount}`);
    return response.data;
  },

  getFeesByStatus: async (status: string): Promise<Fee[]> => {
    const response = await api.get(`/fees/status/${status}`);
    return response.data;
  },

  getFeesBySemester: async (semester: string, academicYear: string): Promise<Fee[]> => {
    const response = await api.get(`/fees/semester/${semester}/year/${academicYear}`);
    return response.data;
  },

  getStudentFeesSummary: async (studentId: number): Promise<FeesSummary> => {
    const response = await api.get(`/fees/student/${studentId}/total`);
    return response.data;
  },

  getPendingFeesByStudent: async (studentId: number): Promise<Fee[]> => {
    const response = await api.get(`/fees/student/${studentId}/pending`);
    return response.data;
  }
};

File: demo/src/main/java/com/example/demo/service/FeeService.java
package com.example.demo.service;

import com.example.demo.dto.FeeRequest;
import com.example.demo.entity.Fee;
import com.example.demo.entity.User;
import com.example.demo.repository.FeeRepository;
import com.example.demo.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
public class FeeService {

    @Autowired
    private FeeRepository feeRepository;

    @Autowired
    private UserRepository userRepository;

    public Fee createFee(FeeRequest feeRequest) {
        Optional<User> student = userRepository.findById(feeRequest.getStudentId());
        
        if (student.isEmpty() || student.get().getRole() != User.Role.STUDENT) {
            throw new RuntimeException("Student not found");
        }

        Fee fee = new Fee(
                student.get(),
                feeRequest.getFeeType(),
                feeRequest.getAmount(),
                feeRequest.getSemester(),
                feeRequest.getAcademicYear()
        );

        fee.setDueDate(feeRequest.getDueDate());
        fee.setDescription(feeRequest.getDescription());

        return feeRepository.save(fee);
    }

    public List<Fee> getStudentFees(Long studentId) {
        return feeRepository.findByStudentId(studentId);
    }

    public List<Fee> getAllFees() {
        return feeRepository.findAll();
    }

    public Optional<Fee> getFeeById(Long id) {
        return feeRepository.findById(id);
    }

    public Fee payFee(Long feeId, BigDecimal amount) {
        Fee fee = feeRepository.findById(feeId)
                .orElseThrow(() -> new RuntimeException("Fee not found"));

        BigDecimal newPaidAmount = fee.getPaidAmount().add(amount);
        
        if (newPaidAmount.compareTo(fee.getAmount()) > 0) {
            throw new RuntimeException("Payment amount exceeds total fee amount");
        }

        fee.setPaidAmount(newPaidAmount);
        fee.setPaidDate(LocalDateTime.now());

        // Update payment status
        if (newPaidAmount.equals(fee.getAmount())) {
            fee.setPaymentStatus(Fee.PaymentStatus.PAID);
        } else if (newPaidAmount.compareTo(BigDecimal.ZERO) > 0) {
            fee.setPaymentStatus(Fee.PaymentStatus.PARTIAL);
        }

        return feeRepository.save(fee);
    }

    public List<Fee> getFeesByStatus(Fee.PaymentStatus status) {
        return feeRepository.findByPaymentStatus(status);
    }

    public List<Fee> getFeesBySemester(String semester, String academicYear) {
        return feeRepository.findBySemesterAndAcademicYear(semester, academicYear);
    }

    public BigDecimal getTotalFeesByStudent(Long studentId) {
        BigDecimal total = feeRepository.findTotalAmountByStudentId(studentId);
        return total != null ? total : BigDecimal.ZERO;
    }

    public BigDecimal getTotalPaidByStudent(Long studentId) {
        BigDecimal total = feeRepository.findTotalPaidAmountByStudentId(studentId);
        return total != null ? total : BigDecimal.ZERO;
    }

    public BigDecimal getOutstandingFeesByStudent(Long studentId) {
        BigDecimal totalFees = getTotalFeesByStudent(studentId);
        BigDecimal totalPaid = getTotalPaidByStudent(studentId);
        return totalFees.subtract(totalPaid);
    }

    public List<Fee> getPendingFeesByStudent(Long studentId) {
        return feeRepository.findByStudentIdAndPaymentStatus(studentId, Fee.PaymentStatus.PENDING);
    }
}